packages/indexing/src/database-index-api.ts

```diff
@@ -115,49 +144,81 @@ export abstract class DatabaseIndexApi<DateType = Date | number> {
+        indexedModels[id] = { streamID: modelData.model, indices: modelIndices }
+        // Add indices to create
+        createIndices[id] = modelIndices
+      } else {
+        const existingIndices = toIndicesRecord(indexedModels[id]?.indices)
+        const addedIndices: IndicesRecord = {}
+        for (const [key, index] of Object.entries(modelIndicesRecord)) {
+          if (existingIndices[key] == null) {
+            addedIndices[key] = index
+          }
+        }
+        // Update indexed models record
+        const allIndices = Object.values({ ...existingIndices, ...addedIndices })
+        indexedModels[id] = { streamID: modelData.model, indices: allIndices }
+        // Update indices to add record only if there is any missing index
+        createIndices[id] = Object.values(addedIndices)
       }
+
+      modelsToIndex.push({ ...modelData, indices: indexedModels[id]?.indices })
     }
+
+    // Ensure all sync checks are done before updating DB
+    await Promise.all(checkSyncPromises)
+    // Index models with possibly new indices to create
+    await this.indexModelsInDatabase(modelsToIndex, createIndices)
+    // Set updated indexed models record once DB has been updated
+    this.indexedModelsRecord = indexedModels
   }
 
-  private async indexModelsInDatabase(models: Array<IndexModelArgs>): Promise<void> {
+  private async indexModelsInDatabase(
+    models: Array<IndexModelArgs>,
+    createIndices: CreateIndicesRecord
+  ): Promise<void> {
     if (models.length === 0) return
-    await this.tablesManager.initMidTables(models)
-    await this.tablesManager.verifyTables(models)
+    await this.tablesManager.initMidTables(models, createIndices)
 
     await this.dbConnection.transaction(async (trx) => {
       // : CDB-1866 - populate the updated_by field properly when auth is implemented
-      await trx(INDEXED_MODEL_CONFIG_TABLE_NAME)
-        .insert(
-          models.map((indexModelArgs) => {
-            return {
-              model: indexModelArgs.model.toString(),
-              ...(indexModelArgs.indices && { indices: JSON.stringify(indexModelArgs.indices) }),
-              is_indexed: true,
-              updated_by: '0', // TODO: FIXME: CDB-1866 - <FIXME: PUT ADMIN DID WHEN AUTH IS IMPLEMENTED>',
-              updated_at: this.now(),
-            }
-          })
-        )
-        .onConflict('model')
-        .merge({
-          updated_at: this.now(),
+      const toInsert = models.map(async (indexModelArgs) => {
+        const data = {
+          model: indexModelArgs.model.toString(),
+          ...(indexModelArgs.indices && { indices: JSON.stringify(indexModelArgs.indices) }),
           is_indexed: true,
           updated_by: '0', // TODO: FIXME: CDB-1866 - <FIXME: PUT ADMIN DID WHEN AUTH IS IMPLEMENTED>',
-        })
+          updated_at: this.now(),
+        }
+        return await trx(INDEXED_MODEL_CONFIG_TABLE_NAME)
+          .insert(data)
+          .onConflict('model')
```

Comment:
constant for this string?

Reply:
It's based on the structure above so I think it's clear as-is?
It's not logic I changed, only moved around.